providers/apis/uniswapv3/fetcher.go

```diff
@@ -0,0 +1,269 @@
+package uniswapv3
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/skip-mev/slinky/oracle/config"
+	"github.com/skip-mev/slinky/oracle/types"
+	uniswappool "github.com/skip-mev/slinky/providers/apis/uniswapv3/pool"
+	"github.com/skip-mev/slinky/providers/base/api/metrics"
+	providertypes "github.com/skip-mev/slinky/providers/types"
+	mmtypes "github.com/skip-mev/slinky/x/marketmap/types"
+)
+
+var _ types.PriceAPIFetcher = (*PriceFetcher)(nil)
+
+// UniswapV3PriceFetcher is the Uniswap V3 price fetcher. This fetcher is responsible for
+// querying Uniswap V3 pool contracts and returning the price of a given ticker. The price is
+// derived from the slot 0 data of the pool contract.
+//
+// To read more about how the price is calculated, see the Uniswap V3 documentation
+// https://blog.uniswap.org/uniswap-v3-math-primer.
+//
+// We utilize the eth client's BatchCallContext to batch the calls to the ethereum network as
+// this is more performant than making individual calls or the multi call contract:
+// https://docs.chainstack.com/docs/http-batch-request-vs-multicall-contract#performance-comparison.
+type PriceFetcher struct {
+	logger  *zap.Logger
+	metrics metrics.APIMetrics
+	api     config.APIConfig
+
+	// client is the EVM client implementation. This is used to interact with the ethereum network.
+	client EVMClient
+	// abi is the uniswap v3 pool abi. This is used to pack the slot0 call to the pool contract
+	// and parse the result.
+	abi *abi.ABI
+	// payload is the packed slot0 call to the pool contract. Since the slot0 payload is the same
+	// for all pools, we can reuse this payload for all pools.
+	payload []byte
+	// poolCache is a cache of the tickers to pool configs. This is used to avoid unmarshalling
+	// the metadata for each ticker.
+	poolCache map[mmtypes.Ticker]PoolConfig
+}
+
+// NewPriceFetcher returns a new Uniswap V3 price fetcher.
+func NewPriceFetcher(
+	logger *zap.Logger,
+	metrics metrics.APIMetrics,
+	api config.APIConfig,
+	client EVMClient,
+) (*PriceFetcher, error) {
+	if logger == nil {
+		return nil, fmt.Errorf("logger cannot be nil")
+	}
+
+	if metrics == nil {
+		return nil, fmt.Errorf("metrics cannot be nil")
+	}
+
+	if api.Name != Name {
+		return nil, fmt.Errorf("expected api config name %s, got %s", Name, api.Name)
+	}
+
+	if !api.Enabled {
+		return nil, fmt.Errorf("api config for %s is not enabled", Name)
+	}
+
+	if err := api.ValidateBasic(); err != nil {
+		return nil, fmt.Errorf("invalid api config: %w", err)
+	}
+
+	abi, err := uniswappool.UniswapMetaData.GetAbi()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get uniswap abi: %w", err)
+	}
+
+	payload, err := abi.Pack("slot0")
```

Comment:
I ref primer for people that want to understand slot0 better.

Reply:
I'm fine also making this a constant, but given that we need the ABI either way i thought it would just be easier to have the payload be generated on start. no strong preference here

Reply:
no, i agree with that. 

I just mean the string literal "slot0" as it is used a few times

Reply:
ohhh

Reply:
Yea lets do that